codeU3.py

```diff
@@ -0,0 +1,222 @@
+        chars = [self.grid[x, y] for (x, y) in route]
+        prefix = ''.join(chars)
+        if self.dict.isPrefix(prefix):
+            return True
+        return False
+
+    def route_word(self, route):
+
+        '''Given a list of tuple coordinates (x,y) encoding route
+        Returns:
+            empty string, if corresponding string is not in the dictionary
+            word, otherwise
+        '''
+
+        if not route:
+            return ""
+        chars = [self.grid[x, y] for (x, y) in route]
+        word = ''.join(chars)
+        if self.dict.isWord(word):
+            return word
+        return ""
+
+    def route_extend(self, route):
+
+        '''Given a route as list of (x,y) indices
+        Returns:
+            all possible routes where 1 step made in one of 8 directions
+            each route is a list of tuple coordinates 
+        '''
+
+        last_x = route[-1][0] #last route coordinate x
+        last_y = route[-1][1] #last route coordinate y
+        cells_set = set(route)
+        route_extended = []
+
+        if last_x + 1 < self.m and (last_x + 1, last_y) not in cells_set: #ensure we don't visit the same cell twice
+            route_extended.append(route + [(last_x + 1, last_y)])
+        if last_x + 1 < self.m and last_y + 1 < self.n and (last_x + 1, last_y + 1) not in cells_set:
+            route_extended.append(route + [(last_x + 1, last_y + 1)])
+        if last_y + 1 < self.n and (last_x, last_y + 1) not in cells_set:
+            route_extended.append(route + [(last_x, last_y + 1)])
+        if last_x - 1 >= 0 and (last_x - 1, last_y) not in cells_set:
+            route_extended.append(route + [(last_x - 1, last_y)])
+        if last_x - 1 >= 0 and last_y + 1 < self.n and (last_x - 1, last_y + 1) not in cells_set:
+            route_extended.append(route+[(last_x - 1, last_y + 1)])
+        if last_x - 1 >= 0 and last_y - 1 >= 0 and (last_x - 1, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x - 1, last_y - 1)])
+        if last_y - 1 >= 0 and (last_x, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x, last_y - 1)])
+        if  last_x + 1 < self.m and last_y - 1 >= 0 and (last_x + 1, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x + 1, last_y - 1)])
+
+        return route_extended
+
+    def search_words(self):
+
+        '''Return all words found in the grid from given dictionary
+        Optimized by stopping searching longer route than maximum length of the dictionary word
+        '''
+
+        all_cells = [[(i, j)] for i in range(self.m) for j in range(self.n)]
+        routes = deque(all_cells)
+        words_found = set()
+        while routes:
+            route = routes.popleft() #queue functionality
+            if self.is_route_prefix(route):
+                word = self.route_word(route)
+                if word:
+                    words_found.add(str(word))

```

Comment:
Indeed, it might be a bit cleaner to do:

```
word = self.route_to_word(route)
if self.dict.is_prefix(word):
  if self.dict.is_word(word):
    ...
```